apiserver/server_test.go

```diff
@@ -512,6 +512,64 @@ func (s *serverSuite) TestAPIHandlerConnectedModel(c *gc.C) {
 	c.Check(handler.ConnectedModel(), gc.Equals, otherState.ModelUUID())
 }
 
+func (s *serverSuite) TestClosesStateFromPool(c *gc.C) {
+	pool := state.NewStatePool(s.State)
+	server := newServer(c, s.State, pool)
+	defer server.Stop()
+
+	otherState := s.Factory.MakeModel(c, nil)
+	defer otherState.Close()
+
+	s.State.StartSync()
+	// This ensures that the model exists for more than one of the
+	// time slices that the watcher uses for coalescing
+	// events. Without it the model appears and disappears quickly
+	// enough that it never generates a change from WatchModels.
+	// Many Bothans died to bring us this information.
+	time.Sleep(coretesting.ShortWait)
+
+	model, err := otherState.Model()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Ensure the model's in the pool but not referenced.
+	st, err := pool.Get(otherState.ModelUUID())
+	c.Assert(err, jc.ErrorIsNil)
+	err = pool.Put(otherState.ModelUUID())
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Make a request for the model API to check it puts
+	// state back into the pool once the connection is closed.
```

Comment:
Well, it couldn't be closed unless it had been put/released, so the test checks it indirectly - the refcount isn't exposed so it can't test it directly (although I could do that if you think it's worthwhile). Originally I had two tests, one that didn't hit the API and one that did, but the former was a subset of the latter so I deleted it.